### PR TITLE
Let MOOSE write MeshMetaData when using `MeshOnlyAction` and `MeshSplitAction` and requesting checkpoint files

### PR DESCRIPTION
**Pull request by arovinelli**
Add a static public member to CheckPoint class to write MeshMetaData in checkpoint files
Call such a method from `MeshOnlyAction` and `MeshSplitAction` to write MeshMetaData to restartable files

ref #16003

### CODE REVIEW

File: framework/include/outputs/Checkpoint.h [hunk 1]
```diff
@@ -54,6 +54,14 @@ class Checkpoint : public FileOutput
    */
   Checkpoint(const InputParameters & parameters);
 
+  /// Method writing MeshMetaData to file. This method is static public so it
+  /// can be called from other classes, thus avoding duplicated code.
+  static void writeMeshMetaData(const processor_id_type pid,
```
Comment (YaqiWang) [inline framework/include/outputs/Checkpoint.h:59]: I feel this function should be a normal function in `MooseApp`. `MooseApp` owns the data. All places calling this function can see `MooseApp`.

Comment (arovinelli) [reply framework/include/outputs/Checkpoint.h:59]: for me it's the same, @permcody suggested to add it in `Checkpoint` as it is part of writing a checkpoint. But you might be right... I'll let you too decide

Comment (YaqiWang) [reply framework/include/outputs/Checkpoint.h:59]: That is also my impression first so pretty sure Cody said that before checking the details of the codes.

**Comment by YaqiWang** [comment]
Do we need to change `SetupRecoverFileBaseAction`?

**Comment by arovinelli** [comment]
> Do we need to change `SetupRecoverFileBaseAction`?

I think we will, but let me figure out reload first

File: framework/src/meshgenerators/FileMeshGenerator.C [hunk 2]
```diff
@@ -83,5 +83,28 @@ FileMeshGenerator::generate()
...
+      std::cerr << "suffix " << suffix << "\n";
+      std::cerr << "map_iter " << map_iter->first << "\n";
+      for (auto const & key : meta_data)
+        std::cerr << key.first << "\n";
+      // if (map_iter->first == "MeshMetaData" && suffix == "_mesh" &&
+      // restartable.readRestartableDataHeader(false, suffix))
+      // {
+
+      // restartable.readRestartableData(meta_data, DataNames());
+      // }
```
Comment (arovinelli) [inline framework/src/meshgenerators/FileMeshGenerator.C:106]: @permcody I definitely need some help here to figure out how to use readRestartableData to read the generated _mesh.rd

Comment (aeslaughter) [reply framework/src/meshgenerators/FileMeshGenerator.C:106]: We do this sort of thing in the stochastic tools module. The action below loads a file.
https://github.com/idaholab/moose/blob/next/modules/stochastic_tools/src/actions/LoadSurrogateDataAction.C

I can help with this if you need it.

File: framework/include/outputs/Checkpoint.h (hunk 1 above)
Comment (aeslaughter) [reply framework/include/outputs/Checkpoint.h:59]: @YaqiWang Is correct registering Mesh meta data is done in the MooseApp. You just need to call MooseApp::registerRestartableData and supply MooseApp::MESH_META_DATA for the last item in the function.

https://github.com/idaholab/moose/blob/8bd5ca2c4d878f8ba48b6ac38a53de0ff348e0df/framework/include/base/MooseApp.h#L521

Comment (aeslaughter) [reply framework/include/outputs/Checkpoint.h:59]: SplitMeshAction is already writing to the data: https://github.com/idaholab/moose/blob/8bd5ca2c4d878f8ba48b6ac38a53de0ff348e0df/framework/src/actions/SplitMeshAction.C#L103

Comment (arovinelli) [reply framework/include/outputs/Checkpoint.h:59]: > SplitMeshAction is already writing to the data:
> 
> https://github.com/idaholab/moose/blob/8bd5ca2c4d878f8ba48b6ac38a53de0ff348e0df/framework/src/actions/SplitMeshAction.C#L103

Yes, this was added by @YaqiWang a few weeks ago via #16218 so we can close this

Comment (aeslaughter) [reply framework/include/outputs/Checkpoint.h:59]: Just noticed how old all this was, the stale bot moved it my notification so I was thinking this was all new. Need to look at dates a bit better. 😆

Comment (arovinelli) [reply framework/include/outputs/Checkpoint.h:59]: @aeslaughter yes this was old, however, I still need your help with #15972.which is the step after this. I don't really get what I'm doing wrong and why I'm not able to relaod data.

**Pull request closed without merging**